neo/SmartContract/StateReader.py

```diff
@@ -286,6 +286,15 @@ def Runtime_Notify(self, engine):
 
         self.notifications.append(args)
 
+        if settings.emit_notify_events_on_sc_execution_error:
+            # emit Notify events even if the SC execution might fail.
+            tx_hash = engine.ScriptContainer.Hash
+            height = Blockchain.Default().Height
+            success = None
+            self.events_to_dispatch.append(NotifyEvent(SmartContractEvent.RUNTIME_NOTIFY, args.State,
+                                                       args.ScriptHash, height, tx_hash,
+                                                       success, engine.testMode))
+
```

Comment:
I've chosen to add event dispatching here over the [alternative location](https://github.com/CityOfZion/neo-python/blob/master/neo/SmartContract/StateReader.py#L220) because this solution makes it so that `Notify` and `Log` events are emitted in order, as opposed to out of order when using the alternative solution. I believe we have to assume people mix `Notify`, `Log` and `Print` statements in their SC code for whatever good reason they might have.

To elaborate take the following SC:
```
def Main(args):
    ctr = 0
    limit = 3
    Notify("Start main")

    while ctr < limit:
        Log(ctr)
        ctr += 1

    Notify("End main")
```

What you want as output is (in order):

> Notify - Start main
> Log - 0
> Log - 1
> Log - 2
> Notify - End Main

What you get in the alternative solution is:
> Log - 0
> Log - 1
> Log - 2
> Notify - Start main
> Notify - End Main